src/appleseed/renderer/kernel/lighting/lighttypes.h

```diff
@@ -71,18 +75,148 @@ class NonPhysicalLightInfo
 class EmittingShape
 {
   public:
-    const AssemblyInstance*     m_assembly_instance;
+    enum ShapeType
+    {
+        TriangleShape = 0
+    };
+
+    static EmittingShape create_triangle_shape(
+        const AssemblyInstance*     assembly_instance,
+        const size_t                object_instance_index,
+        const size_t                primitive_index,
+        const Material*             material,
+        const foundation::Vector3d& v0,
+        const foundation::Vector3d& v1,
+        const foundation::Vector3d& v2,
+        const foundation::Vector3d& n0,
+        const foundation::Vector3d& n1,
+        const foundation::Vector3d& n2,
+        const foundation::Vector3d& geometric_normal);
+
+    ShapeType get_shape_type() const;
+
+    const AssemblyInstance* get_assembly_instance() const;
+
+    size_t get_object_instance_index() const;
+
+    size_t get_primitive_index() const;
+
+    float get_area() const;
+    float get_rcp_area() const;
+
+    float get_shape_prob() const;
+    void set_shape_prob(const float prob);
+
+    const Material* get_material() const;
+
+    void sample_uniform(
+        const foundation::Vector2f& s,
+        const float                 shape_prob,
+        LightSample&                light_sample) const;
+
+    float evaluate_pdf_uniform() const;
+
+    void make_shading_point(
+        ShadingPoint&               shading_point,
+        const foundation::Vector3d& point,
+        const foundation::Vector3d& direction,
+        const foundation::Vector2f& bary,
+        const Intersector&          intersector) const;
+
+    void estimate_average_radiance();
+    float get_average_radiance() const;

```

Comment:
This is my original mistake. The correct term here is I think flux, not radiance.